### PR TITLE
[RSDK-11008] - Custom training script

### PR DESCRIPTION
**Pull request by Rob1in**
This first PR includes:

1. Training script in `src/train.py` with results visualization

2. Full [Hydra](https://hydra.cc/docs/intro/) configuration for experiment configuration
3. A full end-to-end example for training an embedder on mnist dataset
training_curves

4. GenericPairDataset class in src/datasets/generic_dataset.py

- which is a template for implementing pair-based contrastive learning datasets
- Requires only __getitem__ and __len__ methods to be implemented
- Supports the standard format: [(x_i, x_j, label_int)] where label is 1 for positive pairs, 0 for negative pairs

Follow-up:
- documentation (tracked in this [ticket](https://viam.atlassian.net/jira/software/c/projects/RSDK/boards/59?assignee=625878195d1e700069af6d32&selectedIssue=RSDK-11010))
- eval script (very much in progress)
- [Optuna](https://optuna.org/) hyperparameters validation?

### CODE REVIEW

**Review by bhaney — APPROVED** [review]
Thank you for this!
